sdks/go/pkg/beam/core/runtime/harness/hooks.go

```diff
@@ -0,0 +1,112 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Hooks allow runners to tailor execution of the worker to allow for customization
+// of features used by the harness.
+//
+// Examples of customization:
+//
+// gRPC integration
+// session recording
+// profile recording
+//
+// Registration methods for hooks must be called prior to calling beam.Init()
+// Request methods for hooks must be called as part of building the pipeline
+// request for the runner's Execute method.
+
+package harness
+
+import (
+	"context"
+	"fmt"
+
+	fnpb "github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+)
+
+const hooksNamespaceKey = "__hooks:"
+
+// HookKey returns a hook registration key for the specified input.
+func HookKey(in string) string {
+	return hooksNamespaceKey + in
+}
+
+// InitHook is a hook that is called when the harness
+// initializes.
+type InitHook func(context.Context) error
+
+var initHookRegistry = make(map[string]InitHook)
+
+// RegisterInitHook registers a InitHook for the
+// supplied identifier.
+func RegisterInitHook(name string, c InitHook) {
+	initHookRegistry[name] = c
+}
+
+func runInitHooks(ctx context.Context) error {
```

Comment:
The caller currently doesn't check this error. Ditto for other run functions. We should also think about what to do when a hook fail: just ignore/log it? Fail the harness? Remove just the failing hook, but still run the others? If the latter, perhaps we should rather have a single hook struct with callbacks, so that we remove all callbacks if one fails.

Reply:
It's not clear we'd want to remove a hook that failed one time. For example, if I'm writing a bunch of profiles, and the distributed filesystem errors out on one write, do I really want to stop sampling?

Since the hooks have the ability to register and deregister themselves based on options, maybe the right approach is to allow the hook to disable itself?

With regards to error handling, given the lack of a composite error type, and a reasonable behavior for handling an error, the best thing I can think of is to log errors that occur (a decision I may regret), but the outer function doesn't return errors. It ran the hooks, and it's safe to continue.